ethers-ens/src/main/kotlin/io/ethers/ens/EnsProvider.kt

```diff
@@ -0,0 +1,834 @@
+            .getOrElse {
+                RpcResponse.error(Error.AvatarParsing("Error on parsing URI: ${uriRes.resultOrThrow()}", it))
+            }
+    }
+
+    /**
+     * Build ENS name used for reverse address resolution: <address without prefix>.addr.reverse
+     */
+    private fun reverseAddressEnsName(address: Address): String {
+        return "${address.toString().lowercase().substring(2)}.$ENS_DOMAIN_REVERSE_REGISTER"
+    }
+
+    /**
+     * Convert IPFS native URL into HTTPS using [IPFS_GATEWAY], as per
+     * [specification](https://docs-ipfs-tech.ipns.dweb.link/how-to/address-ipfs-on-web/#ipfs-addressing-in-brief).
+     */
+    private fun joinWithIPFSGateway(uri: URI): RpcResponse<URI> {
+        val path = uri.toString().removePrefix("ipfs://").removePrefix("ipfs/")
+        return RpcResponse.result(URL(URL(IPFS_GATEWAY), path).toURI())
+    }
+
+    /**
+     * Possible errors during ens name resolution
+     */
+    sealed class Error : RpcResponse.Error() {
+        /**
+         * Ens name is not valid.
+         */
+        data object EnsNameInvalid : Error()
+
+        /**
+         * Error on ens name normalisation attempt.
+         */
+        data class Normalisation(val cause: Throwable?) : Error() {
+            override fun doThrow() {
+                throw RuntimeException("Normalisation failed: $cause")
+            }
+        }
+
+        // Errors when getting resolver address
+        /**
+         * Error on getting resolver address from registry
+         */
+        data class ResolvingResolver(
+            val registryAddress: Address,
+            val nameHash: String,
+        ) : Error() {
+            override fun doThrow() {
+                throw RuntimeException("Error when getting resolver address from registry> $registryAddress for nameHash: $nameHash.")
+            }
+        }
+
+        /**
+         * Resolver address not found for given nameHash on registry.
+         */
+        data object UnknownResolver : Error()
+
+        // Errors when resolving ENS name with resolver
+        /**
+         * Resolver does not support selector
+         */
+        data class UnsupportedSelector(
+            val resolver: Address,
+            val selector: String,
+        ) : Error() {
+            override fun doThrow() {
+                throw RuntimeException("Resolver: ")
+            }
+        }
+
+        /**
+         * Resolver address resolved ens name to an empty address.
+         */
+        data class UnknownEnsName(
+            val resolverAddr: Address,
+            val nameHash: String,
+        ) : Error() {
+            override fun doThrow() {
+                throw RuntimeException("Resolver: $resolverAddr resolved namehash: $nameHash to an empty address!")
+            }
+        }
+
+        /**
+         * Resolver for ensName exists, but was not able to resolve it.
+         */
+        data class FailedToResolve(val message: String) :
+            Error() {
```

Comment:
This should not be on a new line. We need to tighten the linter checks to auto-correct these kind of cases next time.